src/Uno.UI/UI/Xaml/Input/FocusManager.wasm.cs

```diff
@@ -44,10 +46,30 @@ internal static void ProcessElementFocused(UIElement element)
 				_log.Value.LogDebug($"{nameof(ProcessElementFocused)}() focusedElement={GetFocusedElement()}, element={element}, searching for focusable parent control");
 			}
 
-			foreach (var parent in element.GetParents())
+			foreach (var candidate in element.GetParents().Prepend(element))
 			{
 				// Try to find the first focusable parent and set it as focused, otherwise just keep it for reference (GetFocusedElement())
-				if (parent is TextBlock textBlock && textBlock.IsFocusable)
+
+				// Special handling for RootVisual - which is not focusable on managed side
+				// but is focusable on native side. The purpose of this trick is to allow
+				// us to recognize, that the page was focused by tabbing from the address bar
+				// and focusing the first focusable element on the page instead.
+				if (candidate is RootVisual rootVisual)
+				{
+					var firstFocusable = FocusManager.FindFirstFocusableElement(rootVisual);
+					if (firstFocusable is FrameworkElement frameworkElement)
+					{
+						if (_log.Value.IsEnabled(LogLevel.Debug))
+						{
+							_log.Value.LogDebug(
+								$"Root visual focused - caused by browser keyboard navigation to the page, " +
+								$"moving focus to actual first focusable element - {frameworkElement?.ToString() ?? "[null]"}.");
+						}
+						frameworkElement.Focus(FocusState.Keyboard);
+					}
+					return;
```

Comment:
Is this `return;` needed?

Reply:
Yes. Once we found the right candidate, we should stop. I can probably change it to `break` to match the other if branches